### PR TITLE
performQuery should not return NSManagedObject

### PR DESCRIPTION
**Pull request by crazytonyli**
Related to #21008, which explained why `performQuery` should not return `NSManagedObject`.

## Test Instructions

Go to Reader -> Discover. Tlick the kebab menu on a post. Tap the "Block this Site" button.

## Regression Notes
1. Potential unintended areas of impact
None.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
What's described in the Test Instructions.

3. What automated tests I added (or what prevented me from doing so)
None.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist: N/A

### CODE REVIEW

**Comment by kean** [comment]
There could be assertions in `performQuery` to verify that the return type is not `NSManagedObject`. Ideally, it would be a compile-time or a linter error. I played around with `unavailable` attributes, but I'm not sure there is any easy ways to make it a compile-time error.

**Comment by crazytonyli** [comment]
@kean I have tried make it a compile-time error to return a `NSManagedObject` in the linked PR https://github.com/wordpress-mobile/WordPress-iOS/pull/21008. Swift compile doesn't pickup the unavailable and obsolete function overloads.

I had an idea of implementing a treating warnings as errors mechanism. Maybe via SwiftLint, because Swift compiler doesn't support that yet. I haven't looked into it though.

**Comment by kean** [comment]
> I have tried make it a compile-time error to return a NSManagedObject in the linked PR https://github.com/wordpress-mobile/WordPress-iOS/pull/21008. Swift compile doesn't pickup the unavailable and obsolete function overloads.

ReactiveSwift has been using [unavailable](https://github.com/ReactiveCocoa/ReactiveSwift/blob/master/Sources/UninhabitedTypeGuards.swift) attributes pretty successfully, but they have a very specific setup with their types and generics. I'm not sure how to set it up correctly.

Yeah, I was also thinking SwiftLint. But it could be a runtime assertion, to the very least.

**Review by mokagio — COMMENTED** [review]
@crazytonyli me being lazy, I didn't want to checkout and run this branch. So I wrote some unit tests for `flagSite...` in #21164 

This PR is a refactor, so is not like my tests are useful in showing that this new approach fixes a bug. Still, one would hope they'll still pass with the new code.

I haven't had a chance to try them here as I made my work on `trunk` to ensure a working baseline. What do you think of merging that PR and then rebasing that work?

Aside from that, the code looks good.

**Comment by crazytonyli** [comment]
@crazytonyli Nice! Thanks for adding the tests. I'll rebase once your PR is merged.

**Comment by crazytonyli** [comment]
@mokagio Nice! Thanks for adding the tests. I'll rebase once your PR is merged.

**Review by mokagio — APPROVED** [review]
I was unable to block WordPress.con news, but I suppose that's got to do with the site being "special". I was able to block https://giolodi.com ✅

IMG_7971

Out of scope: I was unable to find a way to unblock it, though. Once blocked, it disappeared from the Reader list and I couldn't find a "blocked" list elsewhere 🤔 That kind of defeat the purpose of having the conditional logic to support unblocking...

**Pull request merged**
